packages/server-wallet/src/objectives/objective-manager.ts

```diff
@@ -0,0 +1,197 @@
+import {Logger} from 'pino';
+import {serializeMessage, Participant, BN, Payload} from '@statechannels/wallet-core';
+import {TransactionOrKnex} from 'objection';
+import {ChannelResult} from '@statechannels/client-api-schema';
+
+import {Bytes32} from '../type-aliases';
+import * as OpenChannelProtocol from '../protocols/open-channel';
+import * as CloseChannelProtocol from '../protocols/close-channel';
+import * as ChannelState from '../protocols/state';
+import {Store} from '../wallet/store';
+import {Channel} from '../models/channel';
+import {LedgerRequest} from '../models/ledger-request';
+import {ChainServiceInterface} from '../chain-service';
+import {Outgoing, ProtocolAction} from '../protocols/actions';
+import {recordFunctionMetrics} from '../metrics';
+
+interface ObjectiveManagerParams {
+  store: Store;
+  logger: Logger;
+  chainService: ChainServiceInterface;
+  timingMetrics: boolean;
+}
+
+// todo: currently duplicated in wallet/index.ts
+type ExecutionResult = {
+  outbox: Outgoing[];
+  channelResults: ChannelResult[];
+  error?: any;
+};
+
+export class ObjectiveManager {
+  private store: Store;
+  private logger: Logger;
+  private chainService: ChainServiceInterface;
+  private timingMetrics: boolean;
+
+  static create(params: ObjectiveManagerParams): ObjectiveManager {
+    return new this(params);
+  }
+
+  private constructor({store, logger, chainService, timingMetrics}: ObjectiveManagerParams) {
+    this.store = store;
+    this.logger = logger;
+    this.chainService = chainService;
+    this.timingMetrics = timingMetrics;
+  }
+
+  async crank(objectiveId: string): Promise<ExecutionResult> {
+    const outbox: Outgoing[] = [];
+    const channelResults: ChannelResult[] = [];
+    let maybeError: any = undefined;
+
+    const objective = await this.store.getObjective(objectiveId);
+
+    let channelToLock;
+
+    if (objective.type === 'OpenChannel' || objective.type === 'CloseChannel')
+      channelToLock = objective.data.targetChannelId;
+    else throw new Error('crankToCompletion: unsupported objective');
+
+    let attemptAnotherProtocolStep = true;
+
+    while (attemptAnotherProtocolStep) {
+      await this.store.lockApp(channelToLock, async tx => {
```

Comment:
I feel like the `lockApp` method might be getting outdated as the paradigm moves forward. In reality, we want to lock all rows related to the `protocolState` that the objective uses, I think. Maybe we can replace `lockApp` with just a simple call to `knex.transaction` or an alias to that e.g.,`store.withinTransaction`. We aren't even using the second argument in the callback anymore (e.g., after `tx` there is a 2nd argument, `channel`, that is omitted on this line 64).

Reply:
Totally agree with this. Was going to tackle this in another PR.

Reply:
Lock app does a really important thing: it synchronizes access to the application.

There needs to be a carefully thought out set of database queries that make updates. There used to be one, which updated a single row and acquired a lock on that row to synchronize that behaviour.

If we are updating multiple rows in multiple relations within a tx, we need to ensure it cannot create a deadlock. If it only needs to update one row that is known in advance, we can avoid locking other rows, but then we have to consider the tx isolation level if there is concurrent updates to other rows read by the transaction, and then we need to expect random Postgres errors

Reply:
I might be wrong, but I'm not sure we actually get that guarantee currently. It looks like we've already moved beyond a world where our updates acquire a single row, do some calculations, and then update that row. The existing `crankToCompletion` function in the wallet appears to lock the channel, not use the value returned, [refetch the channel](https://github.com/statechannels/statechannels/blob/master/packages/server-wallet/src/wallet/index.ts#L696-L700) within the transaction, and in cases [update the objective](https://github.com/statechannels/statechannels/blob/master/packages/server-wallet/src/wallet/index.ts#L746), or [update the ledger tables](https://github.com/statechannels/statechannels/blob/master/packages/server-wallet/src/wallet/index.ts#L746-L750). 

Something I'm aiming for in this set of refactors is to make it clearer what operations are actually happening within each transaction, so we can better reason about whether this is all ok. The plan is to have "atomic" crank operations, that each run in their own transaction, and which make it really easy to see which operations run within that transaction. We can then make a call on whether the existing locking approach is sufficient, or whether we need to go with an approach such as the one outlined in [this doc](https://www.notion.so/Server-wallet-updated-architecture-ad0372e611ac485aa16dc2c4df65a14f) (when we have the bandwidth for a proper discussion).